### ISSUE TITLE
Fix types of Aggregate_Column

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#6866]
Since #6790 we are starting to check types.

This uncovered a sneaky code smell that we relied on to make our life easier when working with `Aggregate_Column`s. While the user was only allowed to identify the columns by their name or index (thus making the API accept `Text | Integer`), the internals (due to legacy that used to allow it explicitly) actually resolve these and replace the identifier with a `Column` reference. This used to work because of no checking. Now this raises a type error.

> Ideally, we'd have some way to create a generic type or type family and have a public `Aggregate_Column allowed_column_types=(Text | Integer)` and a private `Resolved_Column = Aggregate_Column allowed_column_types=Column`. But we have no way to represent that abstraction, so we were just "showing" the public API, but abusing the types in the internal implementation. We need to find a better solution now, to keep the API nice.

To mitigate the issue, a quick fix is being added adding `| Column` to the type signatures. But that shouldn't be the case - the public API should only allow `Text | Integer`.

We need to rewrite the methods related to aggregate columns so that we don't abuse the API anymore. Part of this task is to figure out how. Due to lack of powerful abstractions, one way to do that could be to just duplicate the entire `Aggregate_Column` tree and create a `Resolved_Aggregate_Column` type which is a clone with changed types. That will be typesafe but introduces quite a bit of duplication. Some alternative solutions could also be tried out. 

> This is a refactor that should have been done at some point anyway, as this typesystem abuse was a code-smell that was created because we **did use to allow** `Column` as one of the types and later removed it - and to avoid a huge rework when removing it we picked this simple solution - but it wasn't very clean to begin with.

**Comment by jdunkerley** [comment]
While here lets check out why `Group_By` needs Any as well.

**Comment by radeusgd** [comment]
After #12994 we could replace `| Any` with `| Column` as it will work with both in-memory and DB columns now, so the `Any` workaround is no longer needed.

But even the `| Column` is internal and not really part of the public API, so I guess ideally the refactor would instead change the type of `valid_columns` in `Validated_Aggregate_Columns` to some other structure.
